src/kangaroo.rs

```diff
@@ -0,0 +1,192 @@
+use crunchy::unroll;
+use super::{PLEN, KeccakFamily, RHO, PI, Permutation};
+
+const RC: [u64; 12] = [
+	0x000000008000808b,
+	0x800000000000008b,
+	0x8000000000008089,
+	0x8000000000008003,
+	0x8000000000008002,
+	0x8000000000000080,
+	0x000000000000800a,
+	0x800000008000000a,
+	0x8000000080008081,
+	0x8000000000008080,
+	0x0000000080000001,
+	0x8000000080008008,
+];
+
+#[allow(unused_assignments)]
+pub fn keccakf(a: &mut [u64; PLEN]) {
+    for i in 0..12 {
+        let mut array: [u64; 5] = [0; 5];
+
+        // Theta
+        unroll! {
+            for x in 0..5 {
+                unroll! {
+                    for y_count in 0..5 {
+                        let y = y_count * 5;
+                        array[x] ^= a[x + y];
+                    }
+                }
+            }
+        }
+
+        unroll! {
+            for x in 0..5 {
+                unroll! {
+                    for y_count in 0..5 {
+                        let y = y_count * 5;
+                        a[y + x] ^= array[(x + 4) % 5] ^ array[(x + 1) % 5].rotate_left(1);
+                    }
+                }
+            }
+        }
+
+        // Rho and pi
+        let mut last = a[1];
+        unroll! {
+            for x in 0..24 {
+                array[0] = a[PI[x]];
+                a[PI[x]] = last.rotate_left(RHO[x]);
+                last = array[0];
+            }
+        }
+
+        // Chi
+        unroll! {
+            for y_step in 0..5 {
+                let y = y_step * 5;
+
+                unroll! {
+                    for x in 0..5 {
+                        array[x] = a[y + x];
+                    }
+                }
+
+                unroll! {
+                    for x in 0..5 {
+                        a[y + x] = array[x] ^ ((!array[(x + 1) % 5]) & (array[(x + 2) % 5]));
+                    }
+                }
+            }
+        };
+
+        // Iota
+        a[0] ^= RC[i];
+    }
+}
+
+struct Reduced;
+
+impl Permutation for Reduced {
+    #[inline]
+    fn execute(buffer: &mut [u64; PLEN]) {
+        keccakf(buffer);
+    }
+}
+
+/// KangarooTwelve's length encoding.
+struct EncodedLen {
+    offset: usize,
+    buffer: [u8; 9],
+}
+
+impl EncodedLen {
+    fn new(len: usize) -> Self {
+        let len_view = (len as u64).to_be_bytes();
```

Comment:
Maybe extract `8` to a constant?

Reply:
I don't think it makes sense... the only reason I used 8 is because it simplifies my implementation of this algorithm